### PR TITLE
`groups` Schema

### PR DESCRIPTION
**Pull request by ridhishjain**
**Do not merge, discussion required**
Further addition to #37 

API endpoints:

* **`POST /groups/create_group`** **Admin only**: Frontend would send group_name, confidential and usernames ( a JSON array of usernames to be added in that group ). The array may be empty.

* **`DELETE /groups/remove_group`** **Admin only**: Frontend would send group_id.

* **`POST groups/:group_id/add_branch`** **Moderator only**: receive a branch code and a department code. Users who are having those branch code and department code and not present in the group already, should be added to the group

* **`POST groups/:group_id/remove_branch`** **Moderator only**: receive a branch code and a department code. Users who are having those branch code and department code and not present in the group already, should be removed from the group

* **`POST groups/:group_id/add_members`** **Moderator only**: receive a JSON array of usernames to be added in the group with id as group_id.

* **`DELETE groups/:group_id/remove_members`** **Moderator only**: receive a JSON array of usernames to be removed in the group with id as group_id.

* **`POST groups/:group_id/add_moderator`** **Moderator only**: receive username who should be made an admin ( A group admin can create more group admins or remove group admins, except the creator of the group)

* **`POST groups/:group_id/remove_moderator`** **Moderator only**: receive a username who should be removed from admin

* **`POST groups/:group_id/update_group_name`** **Moderator only**: receive a new group_name, which should be updated.

* **`GET /groups`**: all non-confidential groups in which the user is present.

* **`GET /groups/moderator_groups`** **Moderator only**: all (confidential and non-confidential) the groups in which the user is present as a moderator. This should contain group_name, creator and member_count of the groups.

* **`GET /groups/:group_id`**: specific group (must be non-confidential for normal user, doesn't matter if moderator) with group_name, creator and a list of members of the group. Each member object have the fields - fullName, username, isGroupModerator, profileImg.

Changes are welcome

### CODE REVIEW

File: server/schema/groups/createGroup.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+const schema = {
+  required: ['group_name', 'confidential', 'creator', 'members'],
```
Comment (cjchirag7) [inline server/schema/groups/createGroup.js:2]: creator needs not be send. We can decode the username from token itself.

File: server/schema/groups/addModerator.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+const schema = {
+  required: ['group_name', 'group_id', 'moderator_username'],
```
Comment (cjchirag7) [inline server/schema/groups/addModerator.js:2]: group_name is not required. group_id will be unique

File: server/schema/groups/getGroup.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+const schema = {
+  required: ['group_name', 'group_id'],
```
Comment (cjchirag7) [inline server/schema/groups/getGroup.js:2]: Again, only id should be required

File: server/schema/groups/removeBranch.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
+const schema = {
+  required: ['group_name', 'group_id', 'department', 'branch', 'year'],
```
Comment (cjchirag7) [inline server/schema/groups/removeBranch.js:2]: Change `branch` to `course` everywhere

**Review by cjchirag7 — CHANGES_REQUESTED** [review]
Please make the above changes. Rest seems fine to me.

File: server/schema/groups/createGroup.js (hunk 1 above)
Comment (ridhishjain) [reply server/schema/groups/createGroup.js]: resolved

File: server/schema/groups/createGroup.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
+const schema = {
+  required: ['group_name', 'confidential', 'members'],
```
Comment (ridhishjain) [inline server/schema/groups/createGroup.js:2]: creator name removed

Comment (cjchirag7) [reply server/schema/groups/createGroup.js:2]: great

File: server/schema/groups/updateGroupName.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+const schema = {
+  required: ['group_name', 'group_id', 'new_group_name'],
```
Comment (cjchirag7) [inline server/schema/groups/updateGroupName.js:2]: Here, only `new_group_name` should be required

**Review by cjchirag7 — CHANGES_REQUESTED** [review]
Just a small change required. Rest LGTM

**Review by cjchirag7 — APPROVED** [review]

**Pull request merged**